examples/wmt/models.py

```diff
@@ -0,0 +1,774 @@
+      x, pad_widths, mode='constant', constant_values=x.dtype.type(0))
+  return padded[:, :-1]
+
+
+def sinusoidal_init(max_len=2048,
+                    min_scale=1.0,
+                    max_scale=10000.0):
+  """1D Sinusoidal Position Embedding Initializer.
+
+  Args:
+      max_len: maximum possible length for the input
+      min_scale: float: minimum frequency-scale in sine grating.
+      max_scale: float: maximum frequency-scale in sine grating.
+
+  Returns:
+      output: init function returning `(1, max_len, d_feature)`
+  """
+
+  def init(key, shape, dtype=np.float32):
+    """Sinusoidal init."""
+    del key, dtype
+    d_feature = shape[-1]
+    pe = np.zeros((max_len, d_feature), dtype=np.float32)
+    position = np.arange(0, max_len)[:, np.newaxis]
+    scale_factor = -np.log(max_scale / min_scale) / (d_feature // 2 - 1)
+    div_term = min_scale * np.exp(np.arange(0, d_feature // 2) * scale_factor)
+    pe[:, :d_feature // 2] = np.sin(position * div_term)
+    pe[:, d_feature // 2: 2 * (d_feature // 2)] = np.cos(position * div_term)
+    pe = pe[np.newaxis, :, :]  # [1, max_len, d_feature]
+    return jnp.array(pe)
+
+  return init
+
+
+class AddPositionEmbs(nn.Module):
+  """Adds (optionally learned) positional embeddings to the inputs."""
+
+  def apply(self,
+            inputs,
+            inputs_positions=None,
+            max_len=512,
+            posemb_init=None,
+            cache=None):
+    """Applies AddPositionEmbs module.
+
+    By default this layer uses a fixed sinusoidal embedding table.  If a
+    learned position embedding is desired, pass an initializer to
+    posemb_init.
+
+    Args:
+      inputs: input data
+      inputs_positions: input position indices for packed sequences.
+      max_len: maximum possible length for the input
+      posemb_init: positional embedding initializer, if None, then use a
+        fixed (non-learned) sinusoidal embedding table.
+      cache: flax attention cache for fast decoding.
+
+    Returns:
+      output: `(bs, timesteps, in_dim)`
+    """
+    assert inputs.ndim == 3, ('Number of dimensions should be 3,'
+                              ' but it is: %d' % inputs.ndim)
+    length = inputs.shape[1]
+    pos_emb_shape = (1, max_len, inputs.shape[-1])
+    if posemb_init is None:
+      # Use a fixed (non-learned) sinusoidal position embedding.
+      pos_embedding = sinusoidal_init(
+          max_len=max_len)(None, pos_emb_shape, None)
+    else:
+      pos_embedding = self.param('pos_embedding', pos_emb_shape, posemb_init)
+    pe = pos_embedding[:, :length, :]
+    # We abuse the same attention Cache mechanism to run positional embeddings
+    # in fast predict mode. We could use state variables instead, but this
+    # simplifies invocation with a single top-level cache context manager.
+    # We only use the cache's position index for tracking decoding position.
+    if cache:
+      if self.is_initializing():
+        cache.store(lambda: (4, (1, 1)))
```

Comment:
cache for a self-attention layer is initialized by (num_dimensions, (num_heads, feature_depth_per_head)) - so these numbers are the smallest possible since I don't actually use the key+value arrays that the cache generates - _just_ the index scalar that the cache also produces.   This abuse of the cache mechanism is confusing, and I have a future PR that does all of this using normal neural-net "state" instead... but there's a little more work to be done there and a lot of work to change users' code in-place when the change happens as well...

Reply:
Thanks for adding the comment!